config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/BundleEntityBuilder.java

```diff
@@ -111,10 +185,59 @@ private void loadEntityDependencies(String policyNameWithPath, AnnotatedBundle a
                                     return dependency.getName().equals(PathUtils.extractName(e.getKey()));
                                 }
                             }).findFirst();
-                    Map entityMap = annotatedBundle.getEntities(entityClass);
-                    optionalGatewayEntity.ifPresent(entityEntry -> entityMap.put(entityEntry.getKey(), entityEntry.getValue()));
+                    Map<String, ? extends GatewayEntity> entityMap = annotatedBundle.getEntities(entityClass);
+                    optionalGatewayEntity.ifPresent(entry -> insertGatewayEntity(entityMap, entry.getKey(),
+                            entry.getValue(), excludeReusable, excludedReusableEntities));
                 }
             }
+            // Remove policies of excluded Encasses
+            excludedReusableEntities.forEach(ex -> {
+                Map<String, ? extends GatewayEntity> policyEntities =
+                        annotatedBundle.getEntities(entityTypeRegistry.getEntityClass(EntityTypes.POLICY_TYPE));
+                if (ex instanceof Encass) {
+                    removeExcludedReusablePolicy((Policy) policyEntities.get(((Encass) ex).getPolicy()), annotatedBundle);
+                } else if (ex instanceof Policy) {
+                    removeExcludedReusablePolicy((Policy) ex, annotatedBundle);
+                }
+            });
+        }
+    }
+
+    /**
+     * Inserts the Gateway entity into the Entity Map. It skip adding the entity if excludeReusable is TRUE and the
+     * Gateway Entity is Reusable
+     *
+     * @param entityMapToUpdate Entity Map where GatewayEntity needs to be inserted
+     * @param key Gateway entity key to store against
+     * @param gatewayEntity Gateway entity to be inserted
+     * @param excludeReusable Exclude inserting reusable entity
+     */
+    private void insertGatewayEntity(Map entityMapToUpdate, String key, GatewayEntity gatewayEntity,
+                                     boolean excludeReusable, List<GatewayEntity> excludedReusableEntities) {
+        if (excludeReusable && gatewayEntity instanceof AnnotableEntity && ((AnnotableEntity) gatewayEntity).isReusable()) {
+            excludedReusableEntities.add(gatewayEntity);
+            return; // Return without inserting as its reusable entity
+        }
+        entityMapToUpdate.put(key, gatewayEntity);
+    }
+
+    private void removeExcludedReusablePolicy(Policy policy, AnnotatedBundle annotatedBundle) {
+        if (policy != null) {
+            final Set<Dependency> dependencies = policy.getUsedEntities();
+            for (Dependency dependency : dependencies) {
+                Class<? extends GatewayEntity> entityClass = entityTypeRegistry.getEntityClass(dependency.getType());
+                Map<String, ? extends GatewayEntity> allEntitiesOfType = annotatedBundle.getEntities(entityClass);
+                allEntitiesOfType.entrySet().stream()
```

Comment:
we are not using found entity, what is this logic for ?

Reply:
@uppoju When we exclude a Policy, we need to exclude all the dependencies of that policy as well.

Reply:
In the below code we are getting the matching entity but we are not storing the result and using anywhere

allEntitiesOfType.entrySet().stream()..filter(e -> {
                            GatewayEntity gatewayEntity = e.getValue();
                            if (gatewayEntity.getName() != null) {
                                return dependency.getName().equals(gatewayEntity.getName());
                            } else {
                                return dependency.getName().equals(PathUtils.extractName(e.getKey()));
                            }
                        }).findFirst();

Reply:
@nikhilahuja please check above comment

Reply:
@uppoju fixed!